src/validation.cpp

```diff
@@ -2449,8 +2449,17 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
                 //  artificially set the default assumed verified block further back.
                 // The test against the minimum chain work prevents the skipping when denied access to any chain at
                 //  least as good as the expected chain.
+                // The minimumchainwork threshold is typically aligned with the assumevalid block height.
+                // However, if assumevalid points to an older block while minimumchainwork remains high,
+                //  we may find the assumevalid block in our index before accumulating sufficient chainwork.
+                // Therefore, during reindex operations, we bypass the minimumchainwork check since a prior IBD
+                //  may have been interrupted before reaching the required chainwork threshold.
                 fScriptChecks = (GetBlockProofEquivalentTime(*m_chainman.m_best_header, *pindex, *m_chainman.m_best_header, params.GetConsensus()) <= 60 * 60 * 24 * 7 * 2);
             }
+        } else if (m_chainman.m_blockman.IsReindexing()) {
+            // During a reindex, the assumed valid block may not be in the index
```

Comment:
While the change does seem to fix my personal experience of unexpected script validations, I don't yet understand why reindex does this in the first place - can we fix that instead of adding extra conditions to the script validation checks? Seems a bit hacky and only solves part of the problem, doesn't it?

Reply:
Not necessarily, but it bothers me that we (or at least I) don't understand what's causing the problem in the first place, the current solution feels like a work-around to me - which is still better than what we had, but it's also arguably more complicated and it doesn't seem to fix the underlying problem of why reindex is missing headers in the first place. And are we sure it's a reindex-only bug or can it happen for reindex-chainstate or IBD as well? I haven't dug too deep into this, but I want it to be fixed and was wondering if the apparent lack of interest in the reviews is because others also felt it's a workaround. Thanks for working on this.